### PR TITLE
detectors/heroku: add Heroku resource detector

### LINKED ISSUE #8956: [detectors/heroku] Contribute heroku detector from resourcedetectionprocessor
## Problem Statement

`collector-contrib/processor/resourcedetectionprocessor/internal/heroku` implements a resource detector for Heroku dynos. There is currently no equivalent in `go-contrib/detectors/`, which means Heroku detection is only available to collector-contrib pipelines and not to any Go SDK user.

## Proposed Solution

Add a new `detectors/heroku` package implementing the Heroku detector as a `resource.Detector`, ported from the collector-contrib implementation, so that `collector-contrib/processor/resourcedetectionprocessor` can delegate to it and any other Go SDK user can adopt it.

## Prior Art

- Collector-contrib implementation to port from: [`processor/resourcedetectionprocessor/internal/heroku`](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/resourcedetectionprocessor/internal/heroku)
- Tracking issue: open-telemetry/opentelemetry-collector-contrib#48159

## Additional Context

What the collector-contrib detector does:

- Reads Heroku-specific environment variables (`HEROKU_APP_ID`, `HEROKU_APP_NAME`, `HEROKU_DYNO_ID`, `HEROKU_RELEASE_CREATED_AT`, `HEROKU_RELEASE_VERSION`, `HEROKU_SLUG_COMMIT`).
- Emits: `cloud.provider` (heroku), `heroku.app.id`, `service.name`, `service.instance.id`, `service.version`, `heroku.release.creation_timestamp`, `heroku.release.commit`.

### PR DESCRIPTION
**Pull request by ChunWenHsueh**
Port the Collector resourcedetectionprocessor Heroku detector to a standalone Go SDK resource detector. Add tests, docs, version metadata, autodetect registration, CODEOWNERS, versions.yaml, and changelog entry.

Detected attributes:
- cloud.provider
- heroku.app.id
- service.name
- service.instance.id
- service.version
- heroku.release.creation_timestamp
- heroku.release.commit

Fixes #8956.

### CODE REVIEW

**Comment by paulojmdias** [comment]
@ChunWenHsueh, please sign the CLA.

\cc @iblancasa

**Comment by ChunWenHsueh** [comment]
@paulojmdias Done! I've just signed the CLA.

File: detectors/heroku/version.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package heroku // import "go.opentelemetry.io/contrib/detectors/heroku"
+
+// Version is the current release version of the Heroku resource detector.
+func Version() string {
+	return "0.15.0"
```
Comment (paulojmdias) [inline detectors/heroku/version.go]:
```suggestion
	return "0.16.0"
```

File: detectors/autodetect/go.mod [hunk 2]
```diff
@@ -4,23 +4,24 @@ go 1.25.0
...
-	go.opentelemetry.io/contrib/detectors/hetzner v0.16.0
-	go.opentelemetry.io/otel v1.44.0
-	go.opentelemetry.io/otel/sdk v1.44.0
+	go.opentelemetry.io/contrib/detectors/aws/ec2/v2 v2.5.0
+	go.opentelemetry.io/contrib/detectors/aws/ecs v1.43.0
+	go.opentelemetry.io/contrib/detectors/aws/eks v1.43.0
+	go.opentelemetry.io/contrib/detectors/aws/lambda v0.68.0
+	go.opentelemetry.io/contrib/detectors/azure/azurevm v0.15.0
+	go.opentelemetry.io/contrib/detectors/gcp v1.43.0
+	go.opentelemetry.io/contrib/detectors/heroku v0.15.0
```
Comment (paulojmdias) [inline detectors/autodetect/go.mod]:
```suggestion
	go.opentelemetry.io/contrib/detectors/heroku v0.16.0
```

File: detectors/heroku/heroku.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+package heroku // import "go.opentelemetry.io/contrib/detectors/heroku"
+
+import (
+	"context"
+	"os"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/resource"
+	semconv "go.opentelemetry.io/otel/semconv/v1.40.0"
```
Comment (paulojmdias) [inline detectors/heroku/heroku.go]:
```suggestion
	semconv "go.opentelemetry.io/otel/semconv/v1.41.1"
```

**Comment by ChunWenHsueh** [comment]
@paulojmdias  I've rebased this branch to resolve the merge conflicts with main. It should be ready for another look. PTAL!

File: detectors/heroku/heroku.go (hunk 3 above)
Comment (paulojmdias) [reply detectors/heroku/heroku.go]: Sorry, please revert it as semconv pkg was not released yet 🙏

File: CHANGELOG.md [hunk 4]
```diff
@@ -33,7 +34,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Fix header attributes lost when using sub-spans in `go.opentelemetry.io/contrib/instrumentation/net/http/httptrace/otelhttptrace`. (#8797)
 - Validate `encoding` configuration for OTLP HTTP exporters in `go.opentelemetry.io/contrib/otelconf`. (#8772)
 - Remove the custom body wrapper from the request's body after the request is processed to allow body type comparisons with the original type in `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp` and `go.opentelemetry.io/contrib/instrumentation/github.com/gorilla/mux/otelmux`. (#6914)
-- Unknown or empty HTTP methods now report "_OTHER" instead of "GET" across all HTTP instrumentations to align with OpenTelemetry semantic conventions. (#8868)
+- Unknown or empty HTTP methods now report "\_OTHER" instead of "GET" across all HTTP instrumentations to align with OpenTelemetry semantic conventions. (#8868)
```
Comment (paulojmdias) [inline CHANGELOG.md]: This and all the other changes below should be reverted

**Review by paulojmdias — COMMENTED** [review]
Please just revert the semconv version as 1.41.1 is not available yet. My mistake 🙏

File: CHANGELOG.md (hunk 4 above)
Comment (ChunWenHsueh) [reply CHANGELOG.md]: Just pushed a new version. Please take a look. Thanks!

**Comment by ChunWenHsueh** [comment]
@paulojmdias @dashpole Just pushed a new version fixin CI failures. Please take a look. Thanks!

**Review by paulojmdias — APPROVED** [review]

**Comment by paulojmdias** [comment]
@ChunWenHsueh please fix the conflicts

**Comment by ChunWenHsueh** [comment]
@paulojmdias Fixed. Please take a look. Thanks!

**Review by dashpole — APPROVED** [review]

**Comment by dashpole** [comment]
Please address outstanding copilot comments.  Otherwise, this LGTM

**Review by iblancasa — APPROVED** [review]

**Comment by ChunWenHsueh** [comment]
@dashpole Done. I also added a new test case for the empty env values. Please take a look. Thanks!

File: detectors/heroku/heroku.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+package heroku // import "go.opentelemetry.io/contrib/detectors/heroku"
+
+import (
+	"context"
+	"os"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/resource"
+	semconv "go.opentelemetry.io/otel/semconv/v1.41.0"
```
Comment (MrAlias) [inline detectors/heroku/heroku.go:12]: Current `main` only allows `semconv/v1.43.0`, so this import and the matching test import fail the `depguard` lint check. Please rebase and update the new module's OpenTelemetry dependencies, `go.sum`, implementation, and tests together.

File: detectors/heroku/heroku.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+// on. It returns an empty resource and no error when no Heroku attributes are
+// available.
+func (d *ResourceDetector) Detect(context.Context) (*resource.Resource, error) {
+	var attrs []attribute.KeyValue
+
+	if dynoID, ok := os.LookupEnv(envDynoID); ok && dynoID != "" {
+		attrs = append(attrs, semconv.ServiceInstanceID(dynoID))
+	}
+	if v, ok := os.LookupEnv(envAppID); ok && v != "" {
+		attrs = append(attrs, semconv.CloudProviderHeroku, semconv.HerokuAppID(v))
```
Comment (MrAlias) [inline detectors/heroku/heroku.go:72]: `cloud.provider` is only added when `HEROKU_APP_ID` is set. With partial metadata such as `HEROKU_APP_NAME` or `HEROKU_RELEASE_VERSION` alone, this returns Heroku-derived attributes without the `cloud.provider=heroku` required by the [Heroku semantic conventions](https://opentelemetry.io/docs/specs/semconv/resource/cloud-provider/heroku/).

Could we add the provider whenever any supported non-empty Heroku metadata is detected and update `TestDetect_MissingAppID` accordingly?

**Review by MrAlias — COMMENTED** [review]
Overall, this is well scoped, but it is not ready to merge. Please align the new module with current `main` and ensure partial Heroku metadata still produces a semconv-compliant resource.
